src/main/scala/com/fulcrumgenomics/bam/identifyprimers/PrimerMatcher.scala

```diff
@@ -0,0 +1,325 @@
+    require(rec.mapped)
+    if (rec.positiveStrand) {
+      val pos      = rec.unclippedStart
+      val interval = new Interval(rec.refName, math.max(1, pos-slop), pos+slop)
+      positiveStrandDetector.getOverlaps(interval).iterator()
+        .filter(primer => math.abs(primer.start - pos) <= slop)
+    }
+    else {
+      val pos      = rec.unclippedEnd
+      val interval = new Interval(rec.refName, math.max(1, pos-slop), pos+slop)
+      negativeStrandDetector.getOverlaps(interval).iterator()
+        .filter(primer => math.abs(primer.end - pos) <= slop)
+    }
+  }
+
+  /** Returns a [[LocationBasedPrimerMatch]], otherwise [[None]] */
+  def find(rec: SamRecord): Option[LocationBasedPrimerMatch] = if (rec.unmapped) None else {
+    // Find primers that overlap the record, then prioritize by those matching the same strand.  At most one primer will
+    // be returned.  In the future, it may be nice to either pick one randomly, keep them all, or prioritize by # of
+    // mismatches.
+    val primerMatches: Option[Primer] = getOverlaps(rec).toSeq match {
+      case Seq()                       => None
+      case Seq(primer)                 => Some(primer)
+      case _primers                    =>
+        _primers.filter(_.positiveStrand == rec.positiveStrand) match {
+          case Seq()       => None
+          case Seq(primer) => Some(primer)
+          case ssPrimers   => throw new IllegalStateException(s"Found multiple primers on the same strand for $rec\n\t${ssPrimers.mkString("\n\t")}")
+        }
+    }
+
+    primerMatches
+      .flatMap { p => mismatchAlign(rec, p) }
+      .map { m => // Convert to a [[LocationBasedPrimerMatch]]
+        LocationBasedPrimerMatch(primer = m.primer, numMismatches = m.numMismatches)
+      }
+  }
+}
+
+/**
+  *
+  * @param primers the primers against which to match
+  * @param slop the extra # of read bases to include when mapping gapped alignment tasks.
+  * @param maxMismatchRate the maximum per-base mismatch rate to accept a primer match.
+  */
+private[identifyprimers] class UngappedAlignmentBasedPrimerMatcher
+(val primers: Seq[Primer],
+ val slop: Int,
+ val maxMismatchRate: Double,
+ val kmerLength: Option[Int] = None
+ ) extends PrimerMatcherWithKmerFilter with UngappedAlignment with LazyLogging {
+
+  /** Examines all primers to find the best mismatch-based alignment match. */
+  def find(rec: SamRecord): Option[UngappedAlignmentPrimerMatch] = {
+    val alignments = getPrimersForAlignmentTasks(rec).flatMap { p => mismatchAlign(rec, p) }
+    getBestUngappedAlignment(alignments)
+  }
+
+  /** Examines all primers to find the best gapped-alignment-based match. */
+  private[identifyprimers] def getBestUngappedAlignment(alignments: Seq[UngappedAlignmentPrimerMatch]): Option[UngappedAlignmentPrimerMatch] = {
+    if (alignments.isEmpty) None
+    else {
+      alignments.sortBy(_.numMismatches).take(2) match {
```

Comment:
It's unlikely you'd have primers with a 2-fold variation in length, but shouldn't this be by _rate_ not number?  Or perhaps by number of matching bases instead?